cryodrgn/commands/abinit_homo.py

```diff
@@ -572,7 +564,13 @@ def main(args):
     else:
         start_epoch = 0
 
-    data_iterator = DataLoader(data, batch_size=args.batch_size, shuffle=True)
+    data_iterator = DataLoader(
+        data,
+        sampler=BatchSampler(
+            RandomSampler(data), batch_size=args.batch_size, drop_last=False
+        ),
+        batch_size=None,
+    )
```

Comment:
Using:
```
data_iterator = DataLoader(
    data, batch_size=args.batch_size, shuffle=True, drop_last=True
)
```
tells PyTorch to use [automatic batching](batch_size=None,) - wherein it fetches `batch_size` samples from `data` (making `batch_size` independent calls), makes a list out of it, and then makes a Tensor out of that.

If `data` is already written to hand out datapoints at multiple indices at once (as is the case with our `ImageDataset`, which passes these indices to `ImageSource` that then makes the fewest file access that it can), then the documentation mentions we can [disable automatic batching](https://pytorch.org/docs/stable/data.html#disable-automatic-batching). Essentially by these flags we're telling pytorch:

`batch_size=None` -> disable automatic batching
`batch_sampler=None` (the default anyway so it's not in the code) -> disable automatic batching
`sampler=BatchSampler(RandomSampler(data), batch_size=batch_size, drop_last=False)` -> consult the length of `data` to find out the no. of data points `n`, permute `0`-`n` indices (without replacement), batch them into `batch_size` chunks (without dropping the last), and then pass these batched indices to `data`'s `getitem`. 

I've verified that this is the behavior by stepping through the pytorch code in a debugger. The end result is the same (including the random indices that are used), but fewer `getitem` calls are made.